content/client-lib-development-guide/features.textile

```diff
@@ -1002,23 +1006,23 @@ h3(#activation-state-machine). Activation state machine
 **** @(RSH3d2b)@ Otherwise, make an asynchronous DELETE HTTP request to "/push/deviceRegistrations":/rest-api/#delete-device-registration using the local @DeviceDetails@ 's ID. This operation requires "push device authentication":#push-device-authentication.
 **** @(RSH3d2c)@ Either way, when the registration is done, a @Deregistered@ or @DeregistrationFailed@ event should be fired.
 **** @(RSH3d2d)@ Transitions to @WaitingForDeregistration@.
-*** @(RSH3d3)@ On event @GotPushDeviceDetails@ (note that this will only happen on platforms whose push device details, after first set, can change, e. g. GCM's registration token refresh):
+*** @(RSH3d3)@ On event @GotPushDeviceDetails@ (note that this will only happen on platforms whose push device details, after first set, can change, e. g. FCM's registration token refresh):
 **** @(RSH3d3a)@ If a custom @registerCallback@ was provided to @Push#activate@, pass it the local @DeviceDetails@ updated with the push details.
 **** @(RSH3d3b)@ Otherwise, make an asynchronous PATCH HTTP request to "/push/deviceRegistrations/:deviceId":/rest-api/#update-device-registration using the local @DeviceDetails@ 's push details as body (but only the changed fields, as described in "the REST endpoint documentation":/rest-api/#update-device-registration). This operation requires "push device authentication":#push-device-authentication.
-**** @(RSH3d3c)@ Either way, when the registration is done, a @RegistrationUpdated@ or @UpdatingRegistrationFailed@ event should be fired.
-**** @(RSH3d3d)@ Transitions to @WaitingForRegistrationUpdate@.
-** @(RSH3e)@ State @WaitingForRegistrationUpdate@:
+**** @(RSH3d3c)@ Either way, when the registration is done, a @RegistrationSynced@ or @SyncRegistrationFailed@ event should be fired.
+**** @(RSH3d3d)@ Transitions to @WaitingForRegistrationSync@.
+** @(RSH3e)@ State @WaitingForRegistrationSync@:
 *** @(RSH3e1)@ On event @CalledActivate@:
 **** @(RSH3e1a)@ Makes @Push#activate@ return or call its callback with no error.
-**** @(RSH3e1b)@ Transitions to @WaitingForRegistrationUpdate@.
-*** @(RSH3e2)@ On event @RegistrationUpdated@:
+**** @(RSH3e1b)@ Transitions to @WaitingForRegistrationSync@.
+*** @(RSH3e2)@ On event @RegistrationSynced@:
```

Comment:
I realized while implementing this that now the activation callback won't be called if we go through this:

NotActivated -[CalledActivate]-> WaitingForRegistrationSync -[RegistrationSynced]-> WaitingForNewPushDeviceDetails

We need a step here and below RSH3e3 to call the `registerCallback` provided to `Push#activate` if we come here through a CalledActivate event.

Reply:
Fixed at dcf9e196fc35fcdfd06df0b57919061a989cf9c9.